mktemp: add func to expose functionality (for use in nushell)

tskinn opened the pull request:
Tried to follow @tertsdiepraam guidance [here](https://github.com/uutils/coreutils/issues/5465#issuecomment-1783830036)

Link to https://github.com/uutils/coreutils/issues/5465

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -546,7 +547,22 @@ fn exec(dir: &str, prefix: &str, rand: usize, suffix: &str, make_dir: bool) -> U
+}
+
+pub fn mktemp(
+    dir: &str,
```
Would it be possible to make this into a `&Path`?

tskinn replied:
Yeah sounds good

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -441,7 +442,7 @@ pub fn uu_app() -> Command {
 }
 
-pub fn dry_exec(tmpdir: &str, prefix: &str, rand: usize, suffix: &str) -> UResult<()> {
+pub fn dry_exec(tmpdir: &str, prefix: &str, rand: usize, suffix: &str) -> UResult<PathBuf> {
```
Do you happen to know why this is `pub`?

tskinner-oppfi replied:
Nope. It did seem a little odd.

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -546,7 +547,22 @@ fn exec(dir: &str, prefix: &str, rand: usize, suffix: &str, make_dir: bool) -> U
+    Ok(path)
+}
+
+pub fn mktemp(
```
Could you give this a bit of documentation?

tertsdiepraam reviewed (commented):
Exactly how I imagined it! Thanks! I just have a few additional suggestions.

tskinn commented:
So sorry! I should have had a quick try at implementing `mktemp` in nushell before bothering you. 

After doing so I realized I wasn't exposing the stuff I needed. Since there is a little bit of logic around getting the system tmp dir I just rearranged where that was done exactly so we could reuse that in the pub `mkdir` function. None of the actual logic should have changed.

Basically what changed
- removed pub from dry exec
- created a `new` func for Options struct containing the logic for getting the system tmp dir which is reused in the original `from` function
- `exec` functions don't print but return result with path
- `uumain` prints to stdout
- added new `mktemp` function

Now that I'm looking back I can probably cleanup the arguments to the `mktemp` function. Going to do that.

tskinn commented:
Ok @tertsdiepraam let me know what you think. I'm a rust noob so I'm sure theres lots to change. I can get nushell mktemp working with this api at least.

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -141,24 +141,29 @@ struct Options {
             }
             Some(template) => {
-                let tmpdir = if env::var(TMPDIR_ENV_VAR).is_ok() && matches.get_flag(OPT_T) {
+                let tmpdir = if env::var(TMPDIR_ENV_VAR).is_ok() && treat_as_template {
```
Nushell has their own env var handling, so I think that what we expose should not depend on env vars. Basically what I think should happen is that this method should not have been changed, but the `Options` struct should be exposed, so that they can construct it however they want. Just like how we did it in `cp`: https://github.com/uutils/coreutils/blob/615b562b64f2d392975a697fa36748f5c2759f08/src/uu/cp/src/cp.rs#L224

The `mktemp` function would then also take the `Options` struct instead of all the separate parameters.

tertsdiepraam reviewed (commented):
Nice work! I think it should be done slightly differently, but it's very much going in the right direction, see my other comment. I hope my description is clear, if it's not, feel free to ask for clarification!

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -352,17 +353,18 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     // Create the temporary file or directory, or simulate creating it.
     let res = if dry_run {
-        dry_exec(&tmpdir, &prefix, rand, &suffix)
+        dry_exec(Path::new(&tmpdir), &prefix, rand, &suffix)
```
You don't have to fix this here, but could you add a TODO saying that we should get a `PathBuf` from `clap` instead of `String`?

tertsdiepraam reviewed (commented):
Great! Just one last small suggestion

tertsdiepraam commented on src/uu/mktemp/src/mktemp.rs:
```diff
@@ -297,7 +298,7 @@ impl Params {
 
         Ok(Self {
-            directory,
+            directory: directory.into(),
```
Cool that we indeed already get the `PathBuf` from clap! However, we still need to fix this conversion from `String` inbetween, but I'll open an issue for that.

tertsdiepraam commented:
Thanks! Hope to see you on the nushell side to implement that part!
